Library/Homebrew/os/mac/cache_store.rb

```diff
@@ -0,0 +1,207 @@
+require "dbm"
+require "json"
+
+#
+# `DatabaseCache` is a class acting as an interface to a persistent storage
+# mechanism residing in the `HOMEBREW_CACHE`
+#
+class DatabaseCache
+  # Name of the database cache file located at <HOMEBREW_CACHE>/<name>.db
+  #
+  # @return [String]
+  attr_accessor :name
+
+  # Instantiates new `DatabaseCache` object
+  #
+  # @param  [String] name
+  # @return [nil]
+  def initialize(name)
+    @name = name
+  end
+
+  # Memoized `DBM` database object with on-disk database located in the
+  # `HOMEBREW_CACHE`
+  #
+  # @return [DBM] db
+  def db
+    @db ||= DBM.open("#{HOMEBREW_CACHE}/#{name}", 0666, DBM::WRCREAT)
+  end
+
+  # Close the `DBM` database object after usage
+  #
+  # @return [nil]
+  def close
+    db.close
+  end
+end
+
+#
+# `CacheStore` is an abstract base class which provides methods to mutate and
+# fetch data from a persistent storage mechanism
+#
+# @abstract
+#
+class CacheStore
+  # Instantiates a new `CacheStore` class
+  #
+  # @param  [DatabaseCache] database_cache
+  # @return [nil]
+  def initialize(database_cache)
+    @db = database_cache.db
+  end
+
+  # Inserts new values or updates existing cached values to persistent storage
+  # mechanism
+  #
+  # @abstract
+  # @param  [Any]
+  # @return [nil]
+  def update!(*)
+    raise NotImplementedError
+  end
+
+  # Fetches cached values in persistent storage according to the type of data
+  # stored
+  #
+  # @abstract
+  # @param  [Any]
+  # @return [Any]
+  def fetch(*)
+    raise NotImplementedError
+  end
+
+  # Deletes data from the cache based on a condition defined in a concrete class
+  #
+  # @abstract
+  # @return [nil]
+  def flush_cache!
+    raise NotImplementedError
+  end
+
+  protected
+
+  # A class instance providing access to the `DBM` database object
+  #
+  # @return [DBM]
+  attr_reader :db
+end
+
+#
+# `LinkageStore` is a concrete class providing methods to fetch and mutate
+# linkage-specific data used by the `brew linkage` command
+#
+# If the cache hasn't changed, don't do extra processing in `LinkageChecker`.
+# Instead, just fetch the data stored in the cache
+#
+class LinkageStore < CacheStore
+  # Types of dylibs of the form (label -> array)
```

Comment:
Still don't quite understand this. I'd suggest these are symbols, too.

Reply:
These correspond to the types of values in `LinkageChecker` printed by the [`LinkageChecker::display_items`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/os/mac/linkage_checker.rb#L155-L171) function where [`things.is_a? Hash`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/os/mac/linkage_checker.rb#L160).

Reply:
@AndrewMcBurney could this not be inferred purely from type analysis?

Reply:
It could, but I wanted to pass the values aggregated by type into the function, so I could simply call `format_path_values`, and `format_hash_values` without having to do extra processing in the `CacheStore` to transform the parameters into two arrays.

Reply:
How do you like this implementation?

**cache_store.rb**
```ruby
  def update!(values:, array_values: {}, hash_values: {})
    values.each do |key, value|
      if value.is_a? Hash
        hash_values[key] = value
      else
        array_values[key] = value
      end
    end

    database_cache[keg_name] = ruby_hash_to_json_string(
      array_values: format_array_values(array_values),
      hash_values: format_hash_values(hash_values),
    )
  end
```

**linkage_checker.rb**
```ruby

  # Updates data store with package path values
  def store_dylibs!
    store.update!(
      values: {
        system_dylibs: @system_dylibs,
        variable_dylibs: @variable_dylibs,
        broken_dylibs: @broken_dylibs,
        indirect_deps: @indirect_deps,
        undeclared_deps: @undeclared_deps,
        unnecessary_deps: @unnecessary_deps,
        brewed_dylibs: @brewed_dylibs,
        reverse_links: @reverse_links,
      },
    )
  end
```

Reply:
@AndrewMcBurney Looks good except I'd be tempted to use something like https://flushentitypacket.github.io/ruby/2015/03/31/ruby-keyword-arguments-the-double-splat-and-starsnake.html so you don't need a `values:` parameter but instead can just pass e.g. `system_dylibs:` as parameters and iterate through them all. Also, could just make `values` a non-keyword parameter which is always a hash?

Reply:
Ahh, that's pretty cool! Will make the changes! 🙂